### PR TITLE
Add diagnostics sidecar and support for zipkin in deployed apps

### PR DESCRIPTION
**Pull request by rynowak**
Adds a diagnostics sidecar and the ability to push distributed traces to zipkin for deployed applications.

The sidecar uses the same diagnostics sink code we use for local run, so we can easily add support for providers to each one. The one delta is that the diagnostics code can use service discovery, so in local run we have to provide diagnostics with a URI/key/etc, and in deployed applications that can come from service discovery.

From this point it will be really easy for us to add anything push-based for diagnostics. The main blockers have been worked out 👍

### CODE REVIEW

File: docs/recipes/zipkin.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    app.kubernetes.io/name: 'zipkin'
+spec:
+  selector:
+    app.kubernetes.io/name: zipkin
+  type: ClusterIP
+  ports:
+  - name: http
+    protocol: TCP
+    port: 9411
+    targetPort: 9411
```
Comment (rynowak) [inline docs/recipes/zipkin.yaml:39]: Note, I didn't take the step of actually deploying zipkin for you, because it requires databases.

It really seems like there isn't a *standard* or *official* route for zipkin. We might want to pick Jaeger as the recommended dtrace package. They seem to have lot more options and documentation for deployment.

File: src/Microsoft.Tye.Core/CombineStep.cs [hunk 2]
```diff
@@ -49,12 +50,22 @@ public override Task ExecuteAsync(OutputContext output, ApplicationBuilder appli
...
+            foreach (var sidecar in project.Sidecars)
+            {
+                sidecar.Outputs.Add(ComputeBindings(application, sidecar.Dependencies));
+            }
+
+            return Task.CompletedTask;
+        }
 
-            foreach (var o in service.Dependencies)
+        private ComputedBindings ComputeBindings(ApplicationBuilder application, IEnumerable<string> dependencies)
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/CombineStep.cs:63]: Refactored to a method so I could call it from 2 places.

File: src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs [hunk 3]
```diff
@@ -274,6 +274,36 @@ internal static class KubernetesManifestGenerator
...
+
+                foreach (var sidecar in project.Sidecars)
+                {
+                    sidecar.EnvironmentVariables.Add(new EnvironmentVariableBuilder("TMPDIR")
+                    {
+                        Value = "/var/tye/diagnostics",
+                    });
+                }
+            }
+
```
Comment (rynowak) [inline src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs:306]: All of these pieces took a while to figure out. Basically there's a few things in the way if you want to use eventpipe in a sidecar:

- You need eventpipe to find the other process
- You need a way to find out which process you want (alternatively handle all of them)
  - event pipe just gives you PIDs and that's all
  - I took the approach of using `ps -p <pid> -o command=` to output the args of each process
- If you share process space (needed to enumerate processes in another container) then the target process won't be pid=1.

So basically it gets much more complicated once the number of processes is > 2 (you + one other .net process). I wanted to be resilient for the case where someone did a `dotnet run` or something like that.

Open to ideas about how to make this better.

File: src/Microsoft.Tye.Extensions/Zipkin/ZipkinExtension.cs [hunk 4]
```diff
@@ -14,48 +14,73 @@ internal sealed class ZipkinExtension : Extension
...
-                {
-                    context.Output.WriteDebugLine("zipkin service already configured. Skipping...");
-                }
-                else
+                context.Output.WriteDebugLine("zipkin service already configured. Skipping...");
+            }
+            else
+            {
+                context.Output.WriteDebugLine("Injecting zipkin service...");
+                var service = new ContainerServiceBuilder("zipkin", "openzipkin/zipkin")
```
Comment (rynowak) [inline src/Microsoft.Tye.Extensions/Zipkin/ZipkinExtension.cs:24]: Injecting zipkin as a service is useful for deployment, because then it can show up as a binding, and tye already knows how to prompt for it.

Basically we can model diagnostics sinks as *external* services and a lot of things *just work*.

File: src/Microsoft.Tye.Extensions/Zipkin/ZipkinExtension.cs [hunk 5]
```diff
@@ -14,48 +14,73 @@ internal sealed class ZipkinExtension : Extension
...
 
-                        // make zipkin available as a dependency of everything.
-                        if (!s.Dependencies.Contains(service.Name))
-                        {
-                            s.Dependencies.Add(service.Name);
-                        }
+                    // make zipkin available as a dependency of everything.
+                    if (!s.Dependencies.Contains(service.Name))
+                    {
+                        s.Dependencies.Add(service.Name);
```
Comment (rynowak) [inline src/Microsoft.Tye.Extensions/Zipkin/ZipkinExtension.cs:48]: @jkotalik - this worked out nicely 👌 - this way the sidecar only has a binding for zipkin.

File: src/Microsoft.Tye.Extensions/Zipkin/ZipkinExtension.cs [hunk 6]
```diff
@@ -14,48 +14,73 @@ internal sealed class ZipkinExtension : Extension
...
+                }
+            }
+            else if (context.Operation == ExtensionContext.OperationKind.Deploy)
+            {
+                foreach (var project in context.Application.Services.OfType<ProjectServiceBuilder>())
+                {
+                    // Bring your rain boots.
+                    project.RelocateDiagnosticsDomainSockets = true;
+
+                    var sidecar = new SidecarBuilder("tye-diag-agent", "rynowak/tye-diag-agent", "0.1")
```
Comment (rynowak) [inline src/Microsoft.Tye.Extensions/Zipkin/ZipkinExtension.cs:68]: LOL don't @ me

**Comment by davidfowl** [comment]
We really shouldn’t point to your random docker hub image

**Comment by rynowak** [comment]
Yeah, I'll log an issue :)
